lldb/tools/lldb-dap/DAP.cpp

```diff
@@ -1310,103 +1396,161 @@ void DAP::EventThread() {
+          std::lock_guard<std::mutex> guard(dap_instance->modules_mutex);
           for (uint32_t i = 0; i < num_modules; ++i) {
             lldb::SBModule module =
                 lldb::SBTarget::GetModuleAtIndexFromEvent(i, event);
 
             std::optional<protocol::Module> p_module =
-                CreateModule(target, module, remove_module);
+                CreateModule(dap_instance->target, module, remove_module);
             if (!p_module)
               continue;
 
             llvm::StringRef module_id = p_module->id;
 
-            const bool module_exists = modules.contains(module_id);
+            const bool module_exists =
+                dap_instance->modules.contains(module_id);
             if (remove_module && module_exists) {
-              modules.erase(module_id);
-              Send(protocol::Event{
+              dap_instance->modules.erase(module_id);
+              dap_instance->Send(protocol::Event{
                   "module", ModuleEventBody{std::move(p_module).value(),
                                             ModuleEventBody::eReasonRemoved}});
             } else if (module_exists) {
-              Send(protocol::Event{
+              dap_instance->Send(protocol::Event{
                   "module", ModuleEventBody{std::move(p_module).value(),
                                             ModuleEventBody::eReasonChanged}});
             } else if (!remove_module) {
-              modules.insert(module_id);
-              Send(protocol::Event{
+              dap_instance->modules.insert(module_id);
+              dap_instance->Send(protocol::Event{
                   "module", ModuleEventBody{std::move(p_module).value(),
                                             ModuleEventBody::eReasonNew}});
             }
           }
+        } else if (event_mask & lldb::SBTarget::eBroadcastBitNewTargetSpawned) {
+          auto target = lldb::SBTarget::GetTargetFromEvent(event);
+          auto target_index = debugger.GetIndexOfTarget(target);
+
+          // Set the shared debugger for GPU processes
+          DAPSessionManager::GetInstance().SetSharedDebugger(target_index,
+                                                             debugger);
+
+          // We create "attachCommands" that will select the target that already
+          // exists in LLDB. The DAP instance will attach to this already
+          // existing target and the debug session will be ready to go.
+          llvm::json::Object attach_config;
+          llvm::json::Array attach_commands;
+
+          attach_commands.push_back(llvm::formatv("target list").str());
+          attach_commands.push_back(
+              llvm::formatv("target select {0}", target_index).str());
+
+          // If we have a process name, add command to attach to the same
+          // process name
+
+          attach_config.try_emplace("type", "lldb");
+          attach_config.try_emplace("name", "GPU Session");
```

Comment:
This name has to be configurable by the creator of the target. It might be useful to encode a name in the event or in the target itself. For example, in the GPUAction handler where we create the GPU target, we should be able to specify a custom name for the target that the session will pick up. A default name like `Session <target id>` might be okay as a fallback.
This functionality should be not just for GPUs, as mentioned above.